implement durable timeout in system database recv

demetris-manikas opened the pull request:
This PR aims to address #451.
I only implemented durable timeout in the `recv` method for now so as to get your feedback before implementing
the functionality in the `getEvent` method.

qianl15 commented on src/system_database.ts:
```diff
@@ -37,7 +37,7 @@ export interface SystemDatabase {
 
   send<T>(workflowUUID: string, functionID: number, destinationUUID: string, message: T, topic?: string): Promise<void>;
-  recv<T>(workflowUUID: string, functionID: number, topic?: string, timeoutSeconds?: number): Promise<T | null>;
+  recv<T>(workflowUUID: string, functionID: number, topic?: string, timeoutSeconds?: number, timeoutDurable?: boolean): Promise<T | null>;
```
I don't think we need this `timeoutDurable` flag. Because the old behavior (resetting the timer) is not ideal and more like a bug we want to fix.

demetris-manikas replied:
timeoutDurable  removed

qianl15 commented on src/system_database.ts:
```diff
@@ -524,35 +524,47 @@ export class PostgresSystemDatabase implements SystemDatabase {
+      const timeoutMillis = timeoutSeconds * 1000;
+      const timeoutPromise =  timeoutDurable
+        ? new Promise<void>(async (resolve) => {
+            await this.sleepms(workflowUUID, functionID, timeoutMillis);
```
We can't reuse the functionID here for `sleepms` because it'll cause conflicts in the `operation_outputs` table. We need two function IDs here: one for the actual recv operation and the other one for the `sleepms` operation within recv. 

One solution is to modify both [workflow.ts](https://github.com/dbos-inc/dbos-transact/blob/main/src/workflow.ts#L841) and [debug_workflow.ts](https://github.com/dbos-inc/dbos-transact/blob/main/src/debugger/debug_workflow.ts#L365) to call `this.functionIDGetIncrement();` twice to allocate two function IDs for each `recv`.

demetris-manikas replied:
So that would mean to change
```ts
recv<T>(workflowUUID: string, functionID: number, topic?: string, timeoutSeconds?: number): Promise<T | null>;
````
to

 ```ts
recv<T>(workflowUUID: string, functionID: number, topic?: string, timeoutSeconds?: number, timeoutFunctionID): Promise<T | null>;
````

and do 

```ts
    const functionID: number = this.functionIDGetIncrement();
    const timeoutFunctionID: number = this.functionIDGetIncrement();

    await this.#dbosExec.userDatabase.transaction(async (client: UserDatabaseClient) => {
      await this.flushResultBuffer(client);
    }, { isolationLevel: IsolationLevel.ReadCommitted });
    this.resultBuffer.clear();

    return this.#dbosExec.systemDatabase.recv(this.workflowUUID, functionID, topic, timeoutSeconds, timeoutFunctionID);

```

demetris-manikas replied:
Would another solution be to use
 ```ts 
`${workflowUUID}_blahblah`
```
?

demetris-manikas replied:
Ok.
How would you like me to handle the generation of the functionIDs ?

1.
```ts
  recv<T>(workflowUUID: string, functionID: number, timeoutFunctionID: number, topic?: string, timeoutSeconds?: number): Promise<T | null>;
```
code in `workflow.ts` 
```
    const functionID: number = this.functionIDGetIncrement();
    const timeoutFunctionID: number = this.functionIDGetIncrement();
    ...
    return this.#dbosExec.systemDatabase.recv(this.workflowUUID, functionID, timeoutFunctionID, topic, timeoutSeconds);
```

2.
```ts
recv<T>(workflowUUID: string, generateFunctionID: () => number, topic?: string, timeoutSeconds?: number): Promise<T | null>;
```
```ts
async recv<T>(
    workflowUUID: string,
    generateFunctionID: () => number,
    topic?: string,
    timeoutSeconds: number = DBOSExecutor.defaultNotificationTimeoutSec
  ): Promise<T | null> {
    const funcID = generateFunctionID()
    ....
    await this.sleepms(workflowUUID, generateFunctionID(), timeoutSeconds * 1000);
    ....
    await this.recordNotificationOutput(client, workflowUUID, funcID, message);
```

code in `workflow.ts` 
```ts
return this.#dbosExec.systemDatabase.recv(this.workflowUUID, this.functionIDGetIncrement, topic, timeoutSeconds)
```

I prefer the latter since it will generate ids  only when needed in contrast with the first that
will make functionIDs have gaps in the db which may feel awkward.

qianl15 replied:
I prefer the former one because it's deterministic. The issue with a generateFunctionID function is that it's non-deterministic for recovery and debugging.

An example scenario:
If we want to restart a workflow for recovery which contains a completed `recv`, the `sleepms` inside it, and some other downstream functions. Then the function ID during re-execution will be misaligned because the recovery process will skip `sleepms` as it finds the recorded output for `recv`. And the function ID won't be incremented correctly because we skipped `sleepms` completely. Then all downstream functions will have function IDs offset by 1 compared to the original execution.

demetris-manikas replied:
I was expecting that there was a reason for your suggestion so this is how I implemented it.
the implementation for the debug_workflow will follow soon

demetris-manikas commented:
The implementation for the `getEvent` is not clear at all to me since it is called from two places.
The one from `workflow` is straight forward but how should the `dbos_executor` generate the timeoutFunctionID ?
Suggestions?

demetris-manikas commented:
Also I don't see what changes the `debug_workflow` needs. There is no timeout actually in use there

qianl15 commented:
> The implementation for the `getEvent` is not clear at all to me since it is called from two places. The one from `workflow` is straight forward but how should the `dbos_executor` generate the timeoutFunctionID ? Suggestions?

We only need to make sure the sleep is durable for `getEvent` if it's called from `workflow`. When it's called from `dbos_executor` the function ID is always `undefined` and nothing is persisted in the database. So if the implementation would be similar to `recv`, but will fall back to use a normal non-durable `sleepms` if the function ID is undefined.

qianl15 commented:
> Also I don't see what changes the `debug_workflow` needs. There is no timeout actually in use there

In `debug_workflow`, we still need to increment the function ID twice. Otherwise, downstream functions in debug mode will have their function IDs offset by 1.

demetris-manikas commented:
While trying to implement the proposed solution to `getEvent` the system breaks from the generation of the extra functionID.
Any ideas on why is it so?

qianl15 commented:
> While trying to implement the proposed solution to `getEvent` the system breaks from the generation of the extra functionID. Any ideas on why is it so?

Can you provide more information on this?
I think it's also better to put the `getEvent` implementation in a separate PR.

demetris-manikas commented:
OK about the getEvent.
This should be ready then.
One last question.
Reading the code I have the feeling that if the operation times out then no database access should occur.
Is that so?
If yes it could be implemented by rejecting the Promise after the sleep call with a custom OperationTimedOutError
and by adding a catch in the Promice.race to return null if the error is an OperationTimedOutError instead of hitting the db.

qianl15 commented:
> OK about the getEvent. This should be ready then. One last question. Reading the code I have the feeling that if the operation times out then no database access should occur. Is that so? If yes it could be implemented by rejecting the Promise after the sleep call with a custom OperationTimedOutError and by adding a catch in the Promice.race to return null if the error is an OperationTimedOutError instead of hitting the db.

If the operation times out we will still need to record the output as `null`. The reason is that we want to make sure the behavior is deterministic when we retry or debug workflows. If the first execution returned a `null` then the following retries will see the recorded output in the database and skip it.

qianl15 commented on src/system_database.ts:
```diff
@@ -617,26 +628,24 @@ export class PostgresSystemDatabase implements SystemDatabase {
     const initRecvRows = (await this.pool.query<workflow_events>(`SELECT key, value FROM ${DBOSExecutor.systemDBSchemaName}.workflow_events WHERE workflow_uuid=$1 AND key=$2;`, [workflowUUID, key])).rows;
     if (initRecvRows.length === 0) {
-      await received;
+      // Register the key with the global notifications listener.
```
As we discussed earlier, let's move changes to `getEvent` to a separate PR.

qianl15 reviewed (commented):
Thanks @demetris-manikas for your contribution! This PR looks solid in general. Can you add more tests for the durable recv timeout?

Here's an example of how we test durable sleep: https://github.com/dbos-inc/dbos-transact/blob/main/tests/oaoo.test.ts#L135-L145
You could add another test in `oaoo.test` to make sure 1) sleep the expected seconds when we first run it, and 2) we skip the sleep and recv returns the same result if we run it again.

demetris-manikas commented:
I will add the tests during the next days.
It's getting late here..

qianl15 commented:
> Changes of `getEvent` reverted. The output can still be recorded as null avoiding execution of the following query when the operation has timeout.
> but I guess it is not big of a deal. If you don't wish to implement it for any reason no problem on my side. It is a minor improvement and maybe not worth of the extra complexity in the code. Thanks a lot for your guidance.

I see what you mean, but I think let's not add the extra complexity here.

demetris-manikas commented:
Test added. 
Also changed the expected values. 

1500 is arbitrary.

2000 is the sleep value passed so test against it. 

Finally a 1000ms for something that should return almost immediately seems to much so I changed it to 200 which in turn is again big (on my PC tests pass even with a value of 10ms) so maybe 50 would be a proper value.

qianl15 commented on tests/oaoo.test.ts:
```diff
@@ -130,18 +130,37 @@ describe("oaoo-tests", () => {
     const initTime = Date.now();
     await expect(testRuntime.invokeWorkflow(WorkflowOAOO, workflowUUID).sleepWorkflow(2)).resolves.toBeFalsy();
-    expect(Date.now() - initTime).toBeGreaterThanOrEqual(1500);
+    expect(Date.now() - initTime).toBeGreaterThanOrEqual(2000);
```
We picked a smaller number because the `setTimeout` function might be inaccurate. The actual sleep time could be slightly shorter than 2000 as well. So I'd recommend setting it to a larger number like `1950` but not `2000`.

demetris-manikas replied:
My understanding of the situation is that it may take bit longer but never less. I tried with 2010 and the test usually passes. With 2005 it always passes on my PC (I tried it 10 times in  row without a single failure).
But 1950 it is!

qianl15 reviewed (approved):
Looks good! Thank you :)
